pyk/src/pyk/kore/rule.py

```diff
@@ -0,0 +1,287 @@
+"""Parse KORE axioms into rewrite rules.
+
+Based on the [LLVM Backend's implementation](https://github.com/runtimeverification/llvm-backend/blob/d5eab4b0f0e610bc60843ebb482f79c043b92702/lib/ast/pattern_matching.cpp).
+"""
+
+from __future__ import annotations
+
+from abc import ABC
+from dataclasses import dataclass
+from typing import TYPE_CHECKING, final
+
+from .syntax import And, App, Ceil, Equals, EVar, Implies, In, Not, Rewrites, String, Top
+
+if TYPE_CHECKING:
+    from .syntax import Axiom, Definition, Pattern
+
+    Attrs = dict[str, tuple[Pattern, ...]]
+
+
+class Rule(ABC):
+    lhs: Pattern
+    rhs: Pattern
+    req: Pattern | None
+    ens: Pattern | None
+    priority: int
+
+    @staticmethod
+    def from_axiom(axiom: Axiom) -> Rule:
+        if isinstance(axiom.pattern, Rewrites):
+            return RewriteRule.from_axiom(axiom)
+
+        attrs = (attr.symbol for attr in axiom.attrs)
+        if 'simplification' in attrs:
+            return SimpliRule.from_axiom(axiom)
+
+        return FunctionRule.from_axiom(axiom)
+
+    @staticmethod
+    def extract_all(defn: Definition) -> list[Rule]:
+        filter_attrs = [
+            'assoc',
+            'constructor',
+            'functional',
+            'idem',
+            'symbol-overload',
+            'subsort',
+            'unit',
+        ]
```

Comment:
Maybe the code would be easier to read if you added a function to classify rules, separate from "extracting"/internalising them.  Each class has certain invariants (obviously a `Rewrites` head for rewrite rules, and an `Implies`  head for equations - simplification or function equation as per attribute, but within those there is a mandatory shape of LHS and RHS given by the compiler).

Reply:
While typing this up... this "mandatory shape" is actually a _contract_  between frontend and (any) backend that should be specified (in a specification external to the code, not "in `ModuleToKore.java`" :grimacing: ).

Reply:
Strictly classifying rules is as expensive as extracting them, this is just faster way for pre-classifying them.

But the mandatory shape is enforced by the respective `from_axiom` functions, and if there's a regression in `definition.kore`, these functions would fail.

Reply:
I will try to do something about readability though.